File: src/main/java/com/DavidM1A2/AfraidOfTheDark/common/item/ItemTelescope.java
package com.DavidM1A2.afraidofthedark.common.item;

import com.DavidM1A2.afraidofthedark.AfraidOfTheDark;
import com.DavidM1A2.afraidofthedark.client.gui.AOTDGuiHandler;
import com.DavidM1A2.afraidofthedark.common.capabilities.player.research.IAOTDPlayerResearch;
import com.DavidM1A2.afraidofthedark.common.constants.ModCapabilities;
import com.DavidM1A2.afraidofthedark.common.constants.ModResearches;
import com.DavidM1A2.afraidofthedark.common.item.core.AOTDItem;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.ItemStack;
import net.minecraft.util.ActionResult;
import net.minecraft.util.EnumActionResult;
import net.minecraft.util.EnumHand;
import net.minecraft.util.text.TextComponentString;
import net.minecraft.world.World;

/**
 * Class representing the telescope item used to track meteors
 */
public class ItemTelescope extends AOTDItem
{
	/**
	 * Constructor sets up item properties
	 */
	public ItemTelescope()
	{
		super("telescope");
	}

	/**
	 * Called when the player right clicks with the telescope
	 *
	 * @param worldIn The world that the telescope was right clicked in
	 * @param playerIn The player that right clicked the telescope
	 * @param handIn The hand the telescope is in
	 * @return The result of the right click
	 */
	@Override
	public ActionResult<ItemStack> onItemRightClick(World worldIn, EntityPlayer playerIn, EnumHand handIn)
	{
		// Grab the itemstack the player is holding
		ItemStack itemStack = playerIn.getHeldItem(handIn);
		// Grab the player's research
		IAOTDPlayerResearch playerResearch = playerIn.getCapability(ModCapabilities.PLAYER_RESEARCH, null);
		// First process server side only
		if (!worldIn.isRemote)
			// Test if the player can research astronomy 1
			if (playerResearch.canResearch(ModResearches.ASTRONOMY_1))
			{
				// The player must be above y=128 to see the clouds
				if (playerIn.getPosition().getY() <= 128)
					playerIn.sendMessage(new TextComponentString("I can't see anything through these thick clouds. Maybe I could move to a higher elevation."));
				else
				{
					playerResearch.setResearch(ModResearches.ASTRONOMY_1, true);
					playerResearch.sync(playerIn, true);
				}
			}
			else
			{
				playerIn.sendMessage(new TextComponentString("I can't understand what this thing does."));
			}
		// If we're on client side test if we have the proper research, if so show the GUI
		else
		{
			// Catch both cases where the research is finished and when the server sent us an unlock packet but we don't have it yet so we
			// check if we 'can research' it
			if (playerResearch.isResearched(ModResearches.ASTRONOMY_1) || playerResearch.canResearch(ModResearches.ASTRONOMY_1))
				playerIn.openGui(AfraidOfTheDark.INSTANCE, AOTDGuiHandler.TELESCOPE_ID, worldIn, playerIn.getPosition().getX(), playerIn.getPosition().getY(), playerIn.getPosition().getZ());
		}

		return ActionResult.newResult(EnumActionResult.SUCCESS, itemStack);
	}
}

File: src/main/java/com/davidm1a2/afraidofthedark/common/item/ItemTelescope.kt
package com.davidm1a2.afraidofthedark.common.item

import com.davidm1a2.afraidofthedark.AfraidOfTheDark
import com.davidm1a2.afraidofthedark.client.gui.AOTDGuiHandler
import com.davidm1a2.afraidofthedark.common.constants.ModCapabilities
import com.davidm1a2.afraidofthedark.common.constants.ModResearches
import com.davidm1a2.afraidofthedark.common.item.core.AOTDItem
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.ItemStack
import net.minecraft.util.ActionResult
import net.minecraft.util.EnumActionResult
import net.minecraft.util.EnumHand
import net.minecraft.util.text.TextComponentTranslation
import net.minecraft.world.World

/**
 * Class representing the telescope item used to track meteors
 *
 * @constructor sets up item properties
 */
class ItemTelescope : AOTDItem("telescope")
{
    /**
     * Called when the player right clicks with the telescope
     *
     * @param world  The world that the telescope was right clicked in
     * @param player The player that right clicked the telescope
     * @param hand   The hand the telescope is in
     * @return The result of the right click
     */
    override fun onItemRightClick(world: World, player: EntityPlayer, hand: EnumHand): ActionResult<ItemStack>
    {
        // Grab the itemstack the player is holding
        val itemStack = player.getHeldItem(hand)
        // Grab the player's research
        val playerResearch = player.getCapability(ModCapabilities.PLAYER_RESEARCH, null)!!
        // Test if the player is high enough to use the telescope
        val highEnough = player.position.y > 128

        // Start with server side processing
        if (!world.isRemote)
        {
            // If the player can research the research research it
            if (playerResearch.canResearch(ModResearches.ASTRONOMY_1) && highEnough)
            {
                playerResearch.setResearch(ModResearches.ASTRONOMY_1, true)
                playerResearch.sync(player, true)
            }

            // If the research is researched then test if the player is high enough
            if (playerResearch.isResearched(ModResearches.ASTRONOMY_1) || playerResearch.isResearched(ModResearches.ASTRONOMY_1.preRequisite!!))
            {
                // Tell the player that they need to be higher to see through the clouds
                if (!highEnough)
                {
                    player.sendMessage(TextComponentTranslation("aotd.telescope.not_high_enough"))
                }
            }
            else
            {
                player.sendMessage(TextComponentTranslation("aotd.dont_understand"))
            }
        }

        // If we're on client side and have the proper research and the player is above y=128 to see the stars, show the GUI
        // Don't print anything out client side since the server side takes care of that for us
        if (world.isRemote && highEnough)
        {
            if (playerResearch.isResearched(ModResearches.ASTRONOMY_1) || playerResearch.canResearch(ModResearches.ASTRONOMY_1))
            {
                player.openGui(AfraidOfTheDark.INSTANCE, AOTDGuiHandler.TELESCOPE_ID, world, player.position.x, player.position.y, player.position.z)
            }
        }
        return ActionResult.newResult(EnumActionResult.SUCCESS, itemStack)
    }
}